File: solutions/elixir/resistor-color-trio/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @moduledoc """
  Helpful program so that you don't have to remember the values of resistor bands.
  """

  # Better Be Right Or Your Great Big Values Go Wrong
  @values [:black, :brown, :red, :orange, :yellow, :green, :blue, :violet, :grey, :white]

  @doc """
  Calculate the resistance value in ohms from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms | :megaohms | :gigaohms}
  def label(colors), do: value(colors) |> format_label()

  @doc false
  @spec label(number) :: {number, :ohms | :kiloohms | :megaohms | :gigaohms}
  defp format_label(r) when r < 1_000, do: {r, :ohms}
  defp format_label(r) when r < 1_000_000, do: {r / 1_000, :kiloohms}
  defp format_label(r) when r < 1_000_000_000, do: {r / 1_000_000, :megaohms}
  defp format_label(r), do: {r / 1_000_000_000, :gigaohms}

  @doc false
  @spec value(colors :: [atom]) :: number
  defp value([tens, ones, exponent | _]),
    do: (value(tens) * 10 + value(ones)) * Integer.pow(10, value(exponent))

  defp value(color), do: Enum.find_index(@values, &(&1 == color))
end

File: solutions/go/resistor-color-trio/resistor_color_trio.go
package resistorcolortrio

import (
	"fmt"
	"math"
)

// Better Be Right Or Your Great Big Values Go Wrong
var values = map[string]int{
	"black": 0, "brown": 1, "red": 2, "orange": 3, "yellow": 4,
	"green": 5, "blue": 6, "violet": 7, "grey": 8, "white": 9}

var labels = []struct {
	val  int
	text string
}{{1e9, "giga"}, {1e6, "mega"}, {1e3, "kilo"}}

// Label describes the resistance value given the colors of a resistor.
// The label is a string with a resistance value with an unit appended
// (e.g. "33 ohms", "470 kiloohms").
func Label(colors []string) string {
	r := value(colors)
	for _, label := range labels {
		if r >= label.val {
			return fmt.Sprintf("%d %vohms", r/label.val, label.text)
		}
	}
	return fmt.Sprintf("%d ohms", r)
}

// Value should return the resistance value of a resistor with a given colors.
func value(colors []string) int {
	if len(colors) < 3 {
		return -1
	}
	base := values[colors[0]]*10 + values[colors[1]]
	exp := float64(values[colors[2]])
	return base * int(math.Pow(10, exp))
}
